### PR TITLE
feat(edge): MoE pruning and advanced edge deployment (Rhino-bird Project-03, 犀牛鸟官网实战课题三)

### PR DESCRIPTION
**Pull request by skywalker-lt**
# MoE Pruning and Advanced Edge Deployment

<p align="center">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://raw.githubusercontent.com/skywalker-lt/yolo-master-edge/main/assets/edge_work_tree_p03_dark.png">
    YOLO-Master Edge work since v1.1.0; the pruning, Android runtime, Android app and inference service lanes all feed this PR
  </picture>
</p>

> Focus: Deployment, Pruning, Quantization, and Acceleration  
> Related to: #259, #300
> Docker Image: [DockerHub](https://hub.docker.com/repository/docker/skywalker0501/yolomaster-api/general)
> Exported Models, Builds: [yolo-master-edge Releases](https://github.com/skywalker-lt/yolo-master-edge/releases/tag/v1.1.0)
> Raw Logs: [yolo-master-edge Releases](https://github.com/skywalker-lt/yolo-master-edge/releases/tag/v1.1.0)

## 📋 Summary

Submission for **[实战题目三 (模型剪枝与端侧部署实践)](https://opensource.tencent.com/summer-of-code/project/108/practice)** from the official website of Tencent Open Source. This PR adds `project03/`, `scripts/project03/`, `android/` (runtime module and app), `jetson/11_p03_trt_bench.sh`, `tempo-ncnn/`, `coreml_export/export_coreml_p03.py` and four export and quantization scripts to `examples/YOLO-Master-Cross-Platform-Edge-Deployment`. It lives inside the cross-platform example.

The chain is complete end-to-end: diagnosis with the upstream expert tracker, automatic progressive pruning with the load-time repairs the released checkpoints need, TRT fp16 and explicit QDQ INT8 on the pruned models, Jetson Orin Nano Super and Samsung Galaxy S26 (Snapdragon Elite 8 Gen5) on-device numbers, a REST + WebSocket inference server on four backends with a Docker image, and an API-vs-pure-runtime comparison on the full COCO val2017 set. Every number is measured on the device above; and there are only two small places where the requirement's outstanding (优秀) bar is missed.

### Model Diagnosis (模型诊断分析)

- [x] Upstream `MoEPruner` tracker used as is (`scripts/project03/diagnose_moe.py`); per-layer expert hit shares and average routing weights on the validation set
- [x] Utilization histogram per MoE layer and the cross-layer heatmap (`expert_usage_heatmap.png`), Gini per layer
- [x] Simple / medium / complex scene split by ground-truth instance count (thresholds 3 and 8) with per-scene routing plots; finding: the released checkpoints route the same experts regardless of scene (L1 divergence 0.0005 or 0), so the concentration is a property of the trained router, which is what makes the pruning free
- [x] Three models diagnosed (v0.1-N, UoMoE-N, EsMoE-N); EsMoE-N shown to be a soft mixture with every expert active on every token (Gini 0)

### Intelligence / Automatic Pruning (智能剪枝实现)

- [x] Threshold-driven automatic removal, default 0.15, any list of thresholds in one run; batch over checkpoints x datasets x thresholds (`prune_moe.py`)
- [x] Progressive sweep 5% to 30% with the full validation set at every rung, plan selected under a 0.005 mAP50-95 budget; the plan is identical from 5% to 30% because the unused experts are exactly unused (`sweep.csv`, `sweep.png`)
- [x] Params, GFLOPs, mAP50 and mAP50-95 before and after, per rung (`final.json`)
- [x] Three load-time repairs the released checkpoints need, applied and logged: `add_residual` default for the January-2026 COCO weights (0.007 mAP without it), BatchNorm-terminated routers that upstream's projection finder silently skips (rebuilt by a backwards scan), per-expert buffers shrunk with the surgery
- [x] Pruned checkpoints load with plain `YOLO()` and export with the stock exporter

### Inference and Deployment Optimization (推理优化部署)

- [x] ONNX export of the pruned models; TensorRT fp32 / fp16 / implicit INT8 ladders with pin sets, bisect and leave-one-out ablation, DETAILED engine audit (`quantize_trt.py`); explicit Q/DQ calibrate-only INT8 and the QAT attempt (`qat_moe.py`); model-only vs end-to-end latency with mAP re-validation (`measure_latency.py`)
- [x] Jetson Orin Nano native builds and benchmarks for pruned N / S / M / L (`jetson/11_p03_trt_bench.sh`), with the three Orin build issues and their fixes recorded (`results/orin/DEPLOYMENT_LOG.md`)
- [x] Android: ncnn export of the pruned model with the dense rewrite (`tempo-ncnn/export_ncnn_p03.py`), Vulkan configured and measured in the runtime (opt-in with parity fallback; on this graph CPU fp16 beats Vulkan (seg-N 33 vs 50 ms, pruned N stock export 72 vs 47 ms only before the dense rewrite), so CPU fp16 is the default), mixed INT8 with ACIQ calibration (`results/android/NCNN_INT8_RESULTS.md`)
- [x] Android NPU path: ONNX Runtime with the QNN execution provider on the Hexagon HTP, fp16 and A16W8 models (`scripts/export_onnx_dense.py`, `scripts/quantize_onnx_qnn.py`), 100% HTP placement
- [x] Android sources in the PR: `android/runtime` (Kotlin API over the shared C++ core, ncnn with per-model fp16 policy and Vulkan parity fallback, ONNX Runtime + QNN with measured per-model defaults) and `android/app` (Compose app: live camera, photo batch, bench with history and share, settings, thermal and SoC diagnostics); model payloads are release assets, not tracked
- [x] Sparse SAHI: the example runtime's `--slicing sparse` mode (port of upstream's Sparse SAHI plus a dense-tiling variant) runs the pruned models unchanged; pre-existing in this example, referenced not re-claimed
- [x] Core ML export of the pruned model with the dense k equals E MoE rewrite, ANE-resident (`coreml_export/export_coreml_p03.py`)

### End-to-end System Build (端到端系统构建)

- [x] Inference service with image input (`/v1/infer`, JSON / YOLO txt / COCO JSON / annotated JPEG), multi-image (`/v1/infer/batch`), video upload (`/v1/video`, NDJSON per frame) and live stream (`WS /v1/stream`, binary frames in, JSON out, keep-latest backpressure)
- [x] Latency monitoring: per-stage timings (queue, decode, pre, infer, post) in every response, Prometheus `/metrics`, `/v1/stats` rolling p50 / p95 / p99 per model and stage; bounded queues with 503 Retry-After, per-request deadlines (504), graceful drain
- [ ] Adaptive batching: not implemented by design. Every target in this submission is a batch-1 real-time path (phone, Orin, per-request GPU serving); the multi-image endpoint runs independent batch-1 jobs across the worker pool, and four workers give 345 to 480 img/s. it is a deliberate decision instead of an abandoned function. 
- [x] End-to-end latency before vs after optimization (preprocess + inference + postprocess): A100 fp32 15.20 ms to INT8 12.16 ms per frame; L4 fp32 17.2 ms to fp16 11.1 ms; L40S serving 8.4 ms (fp32) to 7.7 ms (fp16) per request; the finding that preprocessing dominates on server hosts (about 8 ms letterbox vs 3 ms inference)

## 📊 Acceptance criteria 

| criterion | pass bar (✅) | outstanding bar (👍)  | result | class |
|---|---|---|---|---|
| Basics | MoEPruner runs, TensorRT export works | automated, batch capable | `diagnose_moe.py` + `prune_moe.py` sweep checkpoints x datasets x thresholds in one run; `quantize_trt.py` builds whole ladders | 👍 |
| Pruning | params -20%, mAP loss < 1% | params -30%, mAP loss < 0.5% | v0.1-N -55.9% at -0.00001 mAP50-95; UoMoE-N -56.7% at +0.00002; worst of 6 cells -0.0012 | 👍 |
| Acceleration | TensorRT fp16 latency -25% | INT8 -40% with < 1% accuracy loss | fp16 -36.5% (A100), -47.3% (L4), -41.9% (Orin Nano); INT8 Q/DQ at N: -0.80 AP but -37.8% on L4, 2.2 points short of the 👍 latency bar; at S: -46.4% (A100) and -50.8% (Orin Nano) at -0.46 AP | fp16 ✅; INT8 👍 met at S scale, missed at N scale on L4 |
| End-to-end Deployment | Android runs, < 200 ms per frame | < 100 ms, live video | On Samsung S26: 30 ms (ncnn CPU fp16), 11 ms (Hexagon NPU); Compose app with live camera, photo batch, bench and settings; Orin Nano 21.9 ms fp16 | 👍 |
| Completeness | deployment docs and benchmarks | Docker image and one-click script | `API_SERVER_RESULTS.md` here, `docs/API.md` and `deploy/README.md` in yolo-master-edge; image `skywalker0501/yolomaster-api:1.2.0` (Bazel + rules_oci, digest pinned base); `deploy/run.sh`, `docker-compose.yml`; 16 API tests | 👍 |

> ⚠️ Note: the current Android app is an **alpha build**. The frontend is not refined and may behave unexpectedly. Test with caution. 

## 🔎 Work and study established beyond the requirements

- **Implicit TensorRT PTQ cannot reach the 1 AP bar on this architecture.** Calibration coverage dominates up to about 2048 train images (+7.6 AP from coverage alone) and then floors at -2.3 AP; the routing gather / expand chain breaks the INT8 path so the expert convs silently run fp16. Explicit per-channel Q/DQ with the sensitive set kept in fp16 (stem pair `model.0` / `model.1`, routers, DFL) is the method that gets to -0.80 AP, and it is portable across GPUs (0.42102 on L4 from the same ONNX).
- **Where the INT8 damage sits**: bisection and leave-one-out ablation show it is distributed and decays with depth; stem plus first downsample carry about 83% of the backbone damage, MoE blocks about zero under implicit quantization. Recovery is sub-additive, which is why pinning a few modules is not enough under implicit PTQ.
- **INT8 does not buy latency at batch 1 on strong GPUs**: every INT8 engine is slower than or equal to fp16 on A100 and L4 (reformats at the INT8 to float boundaries, launch-bound regime). INT8 wins on the Orin Nano at S / M / L (-5.4% to -7.2% vs fp16) and loses at N, so the recipe is per device and per scale, and the pruned fp16 engine is the default deliverable on GPUs.
- **QAT at short budgets is negative**: 3 epochs against frozen amax scales walked the weights 2.4 AP away from their own PTQ start; the retry recipe (unfrozen scales, longer schedule) is recorded instead of a claimed recovery.
- **On the phone the graph matters more than the precision**: the dense export rewrite that removed the MatMul / Permute / Tile glue around area attention and the router broadcast took seg-N from 76 to 33 ms; mixed INT8 on ncnn is kept for size only (it is slower than fp16 on this CPU); the NPU path with 100% HTP placement is the sub-15 ms option.
- **Serving costs about 1 ms**: the C++ server adds 0.8 to 1.5 ms per request over the bare CLI on every backend, with numerically identical detections, and the pruned model's advantage survives into serving (14% on TensorRT, a wash on x86 CPU backends).

## ⛓️‍💥 Toolchain issues recorded & corresponding fixes

| issue | where it happened | handling |
|---|---|---|
| Released COCO checkpoints predate `add_residual`; loaded raw they score 0.007 | pruning, quantization | default filled at load and logged |
| BatchNorm-terminated routers skipped by the upstream projection finder | pruning | backwards scan rebuilds the projection; per-expert buffers shrunk with the surgery |
| `PREFER_PRECISION_CONSTRAINTS` silently drops every fp16 pin once a calibrator is present | TensorRT INT8 | OBEY plus per-layer output types; pins that split fused SiLU groups make OBEY builds infeasible, so survive-and-revert logic |
| Only the DETAILED engine inspector tells you which precision a layer runs in | TensorRT INT8 | audit committed for every engine |
| TensorRT 11 removed the implicit-quantization API | TensorRT INT8 | pinned to 10.13.3.9 for the ladders |
| `torch.onnx.export` segfaults tracing modelopt fake-quant graphs on CUDA | explicit Q/DQ | trace on CPU |
| Ultralytics `.val()` needs the 4-byte-header engine container | eval | written by `quantize_trt.py` |
| Orin Nano: pure fp16 build asserts at optimization level 2 or below (`sm80` shader with no `sm87` base) | Jetson | `--builderOptimizationLevel=3` |
| Orin Nano: ORT-quantized Q/DQ does not parse (asymmetric activations, int32 bias DQ, opset) | Jetson | symmetric activations, no bias quantization, opset 13 |
| Orin Nano 4 GB: builder OOM | Jetson | headless, 8 GB swap, `--memPoolSize=workspace:256 --maxAuxStreams=0` |
| ncnn fp16 arithmetic overflows on router constants (1e-9 / 1e30 / amax fingerprints) | Android | per-model fp16 policy keyed on the constants; mixture models pinned fp32, v0.1 / EsMoE / pruned models fp16-safe |
| ncnn KL calibration produces a dead INT8 model on this graph | Android INT8 | ACIQ calibration, 2048 images (-1.44 AP vs -3.22 for KL) |
| MNN CUDA fp16 returns zero detections (router overflow) | serving | MNN GPU runs fp32 unless fp16 is explicitly requested; recorded as an invalid cell, not hidden |

## 📦 Builds and logs 

Attached to the [yolo-master-edge v1.1.0 release](https://github.com/skywalker-lt/yolo-master-edge/releases/tag/v1.1.0):

- `project03-pruned-weights-t0.10.tar` (429 MB): the pruned checkpoints for every model x dataset cell and the S / M / L scaling runs
- `project03-pruning.tar.gz`, `project03-uomoe-pruning.tar.gz`: raw diagnosis and sweep outputs
- `project03-trt-v01n-engines.tar` (215 MB), `project03-trt-uomoe-engines.tar` (586 MB), `project03-trt-scaling-engines.tar` (1.09 GB), `project03-trt-other-engines.tar` (231 MB), `project03-qat-engines.tar` (1.40 GB): every A100 and L4 engine of the INT8 study with its inspector audit
- `p03-coreml-v2.tar`: the Core ML package of the pruned model
- `s26_bundle.tgz` (196 MB): the phone-side ncnn and ONNX bundles as benchmarked; `npu_bundle.tgz` (529 MB): the QNN-ready fp16 and A16W8 ONNX models
- `yolomaster-edge-android-app-1.1.0.apk` (260 MB): the app as built from `android/` (arm64-v8a and x86_64, 12 bundled model directories, ONNX Runtime 1.29.0 + Qualcomm QNN 2.42.0 runtime libraries, Qualcomm license applies to those); `yolomaster-edge-android-runtime-1.1.0.aar` (219 MB): the runtime module alone
- `api_bench.tgz`: the 23-cell raw comparison (client latencies, server stage timings, txt dumps, parity reports)
- `yolomaster-edge-jetson-orin-1.1.0.tar.gz` and the Linux / Windows / macOS runtime bundles (pre-existing release assets)
- Docker image: `docker.io/skywalker0501/yolomaster-api:1.2.0` (digest `sha256:00e72c118237...`, 5.25 GB, base `nvidia/cuda:12.9.1-cudnn-runtime-ubuntu22.04` pinned by digest)

> ⚠️ Note: the current Android app is an **alpha build**. The frontend is not refined and may behave unexpectedly. Test with caution. 

## 📁 Files added

```
project03/README.md                      the submission report: criteria table, layout, diagnosis, pruning, TensorRT, Android, serving
project03/results/pruning/               PRUNING_RESULTS.md, UOMOE_PRUNING_RESULTS.md; per cell: report.txt, usage.json, scene_usage.json, sweep.csv/png, plan_t0.10.json, final.json, heatmaps
project03/results/quant/                 INT8_RESULTS.md; trt/ (ladders, sensitivity reports, pin sets, bisect, ablation, calibration-coverage sweeps, L4 ladder, e2e latency, scaling); qat/ (Q/DQ and QAT runs, UoMoE variants); plot script
project03/results/orin/                  DEPLOYMENT_LOG.md (Orin Nano builds, issues, ladder)
project03/results/android/               NCNN_INT8_RESULTS.md (fp16 policy, ACIQ INT8, S26 measurements)
project03/results/api/                   API_SERVER_RESULTS.md (L40S, 23 cells)
scripts/project03/                       diagnose_moe.py, prune_moe.py, quantize_trt.py, qat_moe.py, measure_latency.py
jetson/11_p03_trt_bench.sh               Orin Nano engine builds and benchmarks, pruned v0.1 N / S / M / L
tempo-ncnn/                              export_ncnn_p03.py (dense rewrite), bench.cpp, CMakeLists.txt, run_orin.sh
coreml_export/export_coreml_p03.py       Core ML export of the pruned model
android/                                 runtime module (android/runtime, vendored C++ core under src/main/cpp/core) and Compose app (android/app), README with the build and device steps
scripts/export_ncnn_dense.py, export_onnx_dense.py, quantize_onnx_qnn.py, score_device_dumps.sh   dense exports, A16W8 QNN quantization, device-dump scoring
```

Note: The C++ server sources and the iOS app are from earlier updates, not considered as contribution made for this project. Hence not in this PR. Please check #227 and #238 for more details. 

## ✅ Validation

- Diagnosis and pruning, reproduces the 2 / 2 / 2 plan and the -55.9% cut:
  `python scripts/project03/diagnose_moe.py --model YOLO-Master-v0.1-N.pt --data coco.yaml --scene-split --out out/diag` then `python scripts/project03/prune_moe.py --model YOLO-Master-v0.1-N.pt --data coco.yaml --stats out/diag/usage.json --thresholds 0.05,0.10,0.15,0.20,0.30 --final-eval --out out/prune`
- TensorRT ladders and the INT8 deliverable, on an L4-class GPU with TensorRT 10.x and `nvidia-modelopt`:
  `python scripts/project03/quantize_trt.py --model pruned.pt --data coco.yaml --calib-n 1024 --pins head --out out/trt` (add `--bisect` / `--ablate` for the sensitive set), `python scripts/project03/qat_moe.py --model pruned.pt --data coco.yaml --calib-n 1024 --skip-train --out out/qat`, `python scripts/project03/measure_latency.py --engine out/qat/model.engine --data coco.yaml --revalidate`
- Orin Nano: `bash jetson/11_p03_trt_bench.sh n s m l` on the device
- Serving: `docker run --gpus all -p 8080:8080 -v $PWD/cache:/opt/yolomaster/cache/trt skywalker0501/yolomaster-api:1.2.0`, then `curl localhost:8080/readyz` and `curl -X POST 'localhost:8080/v1/infer?model=v01n-trt&conf=0.25' --data-binary @image.jpg`

### CODE REVIEW

**Comment by isLinXu** [comment]
LGTM
